### PR TITLE
Remove deprecated registry API

### PR DESCRIPTION
**Pull request by bart0sh**

### CODE REVIEW

**Comment by bart0sh** [comment]
This one is not as straightforward as the previous one, so @klihub please take a look. I'm not sure replacing `GetRegistry` with `GetDefaultCache` is correct approach.

**Comment by bart0sh** [comment]
@klihub 
> But one thing I already noticed will need updating is our godoc in pkg/cdi/doc.go.

Thank you for pointing this out! I've updated the docs and code comments.

**Comment by bart0sh** [comment]
@elezar @klihub Can you review this PR, please?

File: cmd/cdi/cmd/root.go [hunk 1]
```diff
@@ -35,13 +35,13 @@ var (
 // rootCmd represents the base command when called without any subcommands
 var rootCmd = &cobra.Command{
 	Use:   "cdi",
-	Short: "Inspect and interact with the CDI Registry",
+	Short: "Inspect and interact with the CDI Cache",
```
Comment (elezar) [inline cmd/cdi/cmd/root.go:38]: nit: Should this be `cache` instead of `Cache`?

File: cmd/cdi/cmd/root.go [hunk 2]
```diff
@@ -68,10 +68,11 @@ func initSpecDirs() {
...
 	if len(specDirs) > 0 {
-		cdi.GetRegistry(
-			cdi.WithSpecDirs(specDirs...),
-		)
-		if len(cdi.GetRegistry().GetErrors()) > 0 {
+		cache := cdi.GetDefaultCache()
+		if err := cache.Configure(cdi.WithSpecDirs(specDirs...)); err != nil {
+			fmt.Printf("failed to configure CDI cache: %v\n", err)
+		}
+		if len(cache.GetErrors()) > 0 {
```
Comment (elezar) [inline cmd/cdi/cmd/root.go:75]:
```suggestion
		cache, err := cdi.NewCache(
			cdi.WithSpecDirs(specDirs...),
		)
		if err != nil {
			fmt.Printf("failed to configure CDI cache: %v\n", err)
		}
		if len(cache.GetErrors()) > 0 {
```

File: cmd/cdi/cmd/root.go [hunk 3]
```diff
@@ -68,10 +68,11 @@ func initSpecDirs() {
 	cdi.SetSpecValidator(validate.WithSchema(s))
 
 	if len(specDirs) > 0 {
-		cdi.GetRegistry(
-			cdi.WithSpecDirs(specDirs...),
-		)
-		if len(cdi.GetRegistry().GetErrors()) > 0 {
+		cache := cdi.GetDefaultCache()
+		if err := cache.Configure(cdi.WithSpecDirs(specDirs...)); err != nil {
+			fmt.Printf("failed to configure CDI cache: %v\n", err)
```
Comment (elezar) [inline cmd/cdi/cmd/root.go:73]: Is printing an error sufficient?

File: pkg/cdi/doc.go [hunk 4]
```diff
@@ -35,48 +35,35 @@
...
 // # Device Injection
 //
-// Using the Registry one can inject CDI devices into a container with code
+// Using the Cache one can inject CDI devices into a container with code
 // similar to the following snippet:
 //
-//	import (
-//	    "fmt"
-//	    "strings"
+//		import (
```
Comment (elezar) [inline pkg/cdi/doc.go:45]: Is the whitespace required?

File: pkg/cdi/doc.go [hunk 5]
```diff
@@ -99,40 +86,40 @@
 // CDI Spec cache is up to date before performing device injection.
 // A code snippet similar to the following accmplishes that:
 //
-//	import (
-//	    "fmt"
-//	    "strings"
+//		import (
```
Comment (elezar) [inline pkg/cdi/doc.go:89]: Similar here.

File: pkg/cdi/doc.go [hunk 6]
```diff
@@ -186,13 +173,13 @@
 //
 //	"fmt"
 //	...
-//	"tags.cncf.io/container-device-interface/specs-go"
+//	specs "tags.cncf.io/container-device-interface/specs-go"
```
Comment (elezar) [inline pkg/cdi/doc.go:176]: Is the `specs` alias strictly needed?

File: pkg/cdi/regressions_test.go [hunk 7]
```diff
@@ -148,12 +148,16 @@ func withCDI(t *testing.T, annotations map[string]string, cdiSpecDirs []string)
 			return nil
 		}
 
-		registry := GetRegistry(WithSpecDirs(cdiSpecDirs...))
-		if err = registry.Refresh(); err != nil {
-			t.Logf("CDI registry refresh failed: %v", err)
+		cache := GetDefaultCache()
+		if err := cache.Configure(WithSpecDirs(cdiSpecDirs...)); err != nil {
```
Comment (elezar) [inline pkg/cdi/regressions_test.go:152]:
```suggestion
		cache, err := NewCache(
			WithSpecDires(cdiSpecDirs...),
		)
		if err != nil {
```

**Review by elezar — CHANGES_REQUESTED** [review]
Thanks @bart0sh.

I have some comments regarding docs. I also think that when it makes sense to do so, we should rather instantiate a `cache` instance explicitly instead of relying on the (global) default cache. Especially in cases where we are configuring the spec directories.

**Comment by bart0sh** [comment]
@elezar @klihub All review comments should be addressed now, please take a look again, thanks!

File: pkg/cdi/doc.go (hunk 6 above)
Comment (bart0sh) [reply pkg/cdi/doc.go]: removed

File: pkg/cdi/doc.go (hunk 5 above)
Comment (bart0sh) [reply pkg/cdi/doc.go]: yeah, it's my editor reformatted those. Should be fixed now. Unfortunately with more reformatting.

File: cmd/cdi/cmd/validate.go [hunk 8]
```diff
@@ -26,22 +26,23 @@ import (
...
-were reported by the registry.`,
+of the CDI cache. It exits with an exit status of 1 if any errors
+were reported by the cache.`,
 	Run: func(cmd *cobra.Command, args []string) {
-		cdiErrors := cdi.GetRegistry().GetErrors()
+		cache := cdi.GetDefaultCache()
+		cdiErrors := cache.GetErrors()
 		if len(cdiErrors) == 0 {
-			fmt.Printf("No CDI Registry errors.\n")
+			fmt.Printf("No CDI Cache errors.\n")
```
Comment (klihub) [inline cmd/cdi/cmd/validate.go:41]: Apparently the original code was inconsistent in this regard, but maybe we could fix it and should always use either lowercase `cache` or capitalized `Cache` in help messages and error/info strings in the `cdi` command. This PR already mostly uses lowercase, but this is an exception here.

File: cmd/cdi/cmd/validate.go [hunk 9]
```diff
@@ -26,22 +26,23 @@ import (
...
 			return
 		}
 
-		fmt.Printf("CDI Registry has errors:\n")
+		fmt.Printf("CDI Cache has errors:\n")
```
Comment (klihub) [inline cmd/cdi/cmd/validate.go:45]: nit: ditto lower vs. upper case here...

**Review by klihub — APPROVED** [review]
@bart0sh I have two very small nitpicks about help and info/error message consistency wrt. `cache` vs. `Cache` in messages. The original code was apparently inconsistent in this regard (no wonder, knowing the author), but this PR is already very close two fixing that with two exceptions.

Otherwise LGTM. Not a showstopper though, hence the approval.

File: cmd/cdi/cmd/validate.go (hunk 8 above)
Comment (bart0sh) [reply cmd/cdi/cmd/validate.go]: Thank you for the review! Should be fixed now.

File: cmd/cdi/cmd/validate.go (hunk 9 above)
Comment (bart0sh) [reply cmd/cdi/cmd/validate.go]: done

File: pkg/cdi/doc.go [hunk 10]
```diff
@@ -106,17 +93,17 @@
...
 //
 //	    "tags.cncf.io/container-device-interface/pkg/cdi"
-//	    oci "github.com/opencontainers/runtime-spec/specs-go"
+//	    "github.com/opencontainers/runtime-spec/specs-go"
 //	)
 //
-//	func injectCDIDevices(spec *oci.Spec, devices []string) error {
-//	    registry := cdi.GetRegistry()
+//	func injectCDIDevices(spec *specs.Spec, devices []string) error {
+//		cache := cdi.GetDefaultCache()
```
Comment (elezar) [inline pkg/cdi/doc.go]: nit: indent

**Review by elezar — APPROVED** [review]
I made some changes to the whitespace in doc.go.

Looks good otherwise.

File: pkg/cdi/doc.go (hunk 10 above)
Comment (bart0sh) [reply pkg/cdi/doc.go]: Thank you. Can you merge the PR?

**Pull request merged**